[Chunk Data Pack Pruner]  Add Iterator Creator

zhangchiqing opened the pull request:
This PR replaced the height based block iterator with the generic block iterator, that could work for both height and view based block iteration.

Added creators for creating height based iterator or view based iterator.

zhangchiqing commented on module/block_iterator/creator_test.go:
```diff
@@ -0,0 +1,334 @@
+	blocks = append(blocks, additionalBlocks...)
+
+	// Create another iterator
+	iterator, err = creator.Create()
```
Creator is reused to create new iteration after previous iteration is completed. So when chunk data pack pruner, after pruning to latest block, the creator can create a new iterator to prune to the new latest height.

zhangchiqing commented on module/block_iterator/creator_test.go:
```diff
@@ -0,0 +1,334 @@
+	)
+	require.NoError(t, err)
+
+	newIterator, err := newCreator.Create()
```
This test simulates a restart, which will create a new creator and resume from last checkpointed iterated block and iterate until the latest.

zhangchiqing commented on module/block_iterator.go:
```diff
@@ -58,64 +48,19 @@ type BlockIterator interface {
-	CreateIterator(IterateRange, IterateProgressWriter) (BlockIterator, error)
-}
-
-type IteratorFactory struct {
```
These logics are moved to creator to simplify.

zhangchiqing commented on module/block_iterator/creator.go:
```diff
@@ -0,0 +1,110 @@
+func NewHeightBasedCreator(
+	getBlockIDByHeight func(height uint64) (flow.Identifier, error),
+	progress storage.ConsumerProgress,
+	root *flow.Header,
+	latest func() (*flow.Header, error),
```
Unlike `NewCreator`, the `NewHeightBasedCreator` and `NewViewBasedCreator` take `flow.Header` instead of `uint64`, in order to ensure the same field (`Height` or `View`) is read from the header of the root block and latest block.

peterargue commented on module/block_iterator/creator.go:
```diff
@@ -0,0 +1,110 @@
+func NewCreator(
+	getBlockIDByIndex func(uint64) (blockID flow.Identifier, indexed bool, exception error),
+	progressStorage storage.ConsumerProgress,
+	root uint64,
```
can you add some context about how `root` is used in the godocs

peterargue commented on module/block_iterator/iterator_test.go:
```diff
@@ -32,38 +32,41 @@ func TestIterateHeight(t *testing.T) {
-		require.Empty(t, visited)
 
-		// save the next to iterate height and verify
+		require.Equal(t, len(bs), len(visited))
```
do this before the loop to avoid a panic if `visited` is smaller

zhangchiqing replied:
The reason I do the for loop check is so that if there is one block not visited, the test case can tell which block is not visited, rather than saying the total visited is different.

peterargue commented on module/block_iterator/creator_test.go:
```diff
@@ -0,0 +1,334 @@
+	}
+
+	// Verify no extra blocks were visited
+	require.Equal(t, len(blocks), len(visitedBlocks), "Unexpected number of blocks visited")
```
same comment here about doing this before the loop

peterargue commented on module/block_iterator/creator_test.go:
```diff
@@ -0,0 +1,334 @@
+	}
+
+	// Verify no extra blocks were visited
+	require.Equal(t, len(additionalBlocks), len(visitedBlocks), "Unexpected number of blocks visited")
```
here too

peterargue reviewed (approved):
added a few testing related comments but otherwise looks good

fxamacker reviewed (commented):
The code makes sense but it uses many interfaces and abstractions, so it can be complicated to understand and use.

Maybe we can simplify design and focus on these data types:
- `PersistentIteratorState` loads state, saves state, and creates next `IteratorRange`.
- `IteratorRange` is a value object with iteration range.
- `BlockIterator` iterates given range with given `getBlockIDByIndex`.

For example, it can be simplified to something like:
```Go
type PersistentIteratorState struct{
    ...
}

func NewPersistentIteratorState(
	root *flow.Header,
	latest func() (*flow.Header, error),
) *PersistentIteratorState

func (*PersistentIteratorState) LoadState() (uint64, error) 

func (*PersistentIteratorState) SaveState(uint64) error 

func (*PersistentIteratorState) NextRange() IteratorRange

type IteratorRange struct {
	Start uint64
	End uint64
}

func NewIndexedBlockIterator(
	getBlockIDByIndex func(uint64) (blockID flow.Identifier, indexed bool, excpetion error),
	range IteratorRange,
) BlockIterator

func IterateBlock(BlockIterator, persistentIteratorRange, callback func(flow.BlockID, ...)) {
	defer persistentIteratorRange.SaveState()

	for iterator.HasNext() {
		id, _, err := iterator.Next()
		...
		callback(id, ...)
	}
}
```

Example using the suggested simplified objects:

```Go
persistentIteratorState := NewPersistentIteratorState(root, latest)

iteratorRange := persistentIteratorState.NextRange() 

itertor := NewIndexedBlockIterator(getBlockIDByIndex, iteratorRange) 

IterateBlock(iterator, persistentIteratorState, iteratorCallback)
```

zhangchiqing commented:
@fxamacker good idea to simply the interfaces. I've merged initializer and range into the iteration progress, and also renamed it into the `PersistentIteratorState`. It's more clear now to state is initialized, updated. 

I still need to keep the `Creator`, so that we can have `NewHeightBasedCreator` and `NewViewBasedCreator` built on top of it for different iteration operations.

fxamacker reviewed (approved):
Thanks for updating this PR!  I mostly focused on non-test code and left some suggestions.

zhangchiqing commented on module/block_iterator/state.go:
```diff
@@ -0,0 +1,64 @@
+	return module.IterateRange{
+		Start: next,
+		End:   latest,
+	}, nil
```
Good catch!

I introduced a bool value (hasNext) to indicate whether there is a next range to iterate through. I didn't use an error since I'd like to keep the returned error as exception.
